cpp/oneapi/dal/algo/kmeans_init/test/batch.cpp

```diff
@@ -0,0 +1,118 @@
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/algo/kmeans_init/compute.hpp"
+
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/row_accessor.hpp"
+#include "oneapi/dal/test/engine/fixtures.hpp"
+
+namespace oneapi::dal::kmeans_init::test {
+
+namespace te = dal::test::engine;
+
+template <typename TestType>
+class kmeans_init_batch_test : public te::float_algo_fixture<std::tuple_element_t<0, TestType>> {
+public:
+    using Float = std::tuple_element_t<0, TestType>;
+    using Method = std::tuple_element_t<1, TestType>;
+
+    auto get_descriptor(std::int64_t cluster_count) const {
+        return kmeans_init::descriptor<Float, Method>{ cluster_count };
+    }
+
+    bool not_available_on_device() {
+        constexpr bool is_plus_plus_dense =
+            std::is_same_v<Method, kmeans_init::method::plus_plus_dense>;
+        constexpr bool is_parallel_plus_dense =
+            std::is_same_v<Method, kmeans_init::method::parallel_plus_dense>;
+        return this->get_policy().is_gpu() && (is_plus_plus_dense || is_parallel_plus_dense);
+    }
+
+    constexpr bool is_dense() const {
+        return std::is_same_v<Method, method::dense> ||
+               std::is_same_v<Method, method::random_dense> ||
+               std::is_same_v<Method, method::plus_plus_dense> ||
+               std::is_same_v<Method, method::parallel_plus_dense>;
+    }
+
+    void dense_checks(std::int64_t cluster_count, const table& data) {
+        CAPTURE(cluster_count);
+
+        INFO("create descriptor")
+        const auto desc = get_descriptor(cluster_count);
+
+        INFO("compute");
+        const auto compute_results = compute(desc, data);
+        check_results(cluster_count, data, compute_results.get_centroids());
+    }
+
+    void check_results(std::int64_t cluster_count, const table& data, const table& centroids) {
+        std::int64_t row_count = data.get_row_count();
+        std::int64_t column_count = data.get_column_count();
+        ONEDAL_ASSERT(centroids.get_row_count() == cluster_count);
+        ONEDAL_ASSERT(centroids.get_column_count() == column_count);
+        const auto data_ptr = row_accessor<const float>(data).pull().get_data();
+        const auto centroid_ptr = row_accessor<const float>(centroids).pull().get_data();
```

Comment:
❗ This is wrong use of `dal::array`. The problem is the object returned by `.pull()` is temporal. It's destroyed at the same line. As a result you'll get a dangling pointers to `data` and `centroids`.

Tip: you have read-only access to the array objects itself via overloaded `operator[](std::int64) const`

Reply:
Fixed